### ISSUE TITLE
Where are the tokens saved?

### ISSUE THREAD

**Issue by tumetus** [AzureAD/azure-activedirectory-library-for-cordova#121]
Where does the plugin save the tokens? I am asking because I want to persist the tokens between app/phone restarts and now it seems that the silent login doesn't always work after phone restart. I have tried to look through the code but can't quite get the grasp of it so any insight would be much appreciated!

**Comment by vladimir-kotikov** [comment]
@tumetus, the plugin itself doesn't store tokens - token cache is managed by native SDKs (that run under the hood of the plugin) and in theory should persist betwen app/device restarts. If it is empty after device restart that might be either a bug or some specific case, where this behavior is expected (e.g. brokered authentication on Android, see #117).

If you're seeing this consistently, could you please provide issue details (as per new issue template) so we can investigate this.

**Comment by tumetus** [comment]
Yeah sure! I'll provide more details when I investigate this further.

**Comment by vladimir-kotikov** [comment]
@tumetus, I'm closing this one. Though feel free to ask any questions here or reopen in case if you're experiencing any issues w/ token cache

**Comment by tumetus** [comment]
@vladimir-kotikov I had some time to test this behaviour further. Not sure if this is the right place to explain this or should I open a new issue?

Anyhow this is what I did:
I set my token lifetimes as follows: access token 10min, refresh token 14days.
I logged in with my ad credentials in my cordova app. I waited for 30mins and tried to aquire token again and it was done so silently (no UI shown). So far everything worked good. I waited for another 5hours after this and tried to aquire token again and could not do it silently anymore. This time the UI for ad login was shown. So first time it seems that the refresh tokens were used correctly and on second attempt they were not?

I tested this with iPhone 6s running iOS 10.2.1. I restarted the app between attempts. I did not restart my device.

**Comment by vladimir-kotikov** [comment]
@tumetus, thanks for additional info! We'll try to reproduce this and let you know about results.It'd be also very helpful if you could
- provide an error returned by failed `acquireTokenSilentAsync` call
- record the logs, written by ADAL to xcode console when `acquireTokenSilent` fails to refresh the token.

**Comment by tumetus** [comment]
Thanks for getting back to me @vladimir-kotikov !

[Here](https://gist.github.com/tumetus/9c0118a46664b24a9279f2bcb41b6dee) is the xcode console printout.

It looks like it interpers the refresh token as expired. Is there a way to inspect the refresh_token parameter? So far I have only managed to inspect the access token.

**Comment by vladimir-kotikov** [comment]
@tumetus, that explains why you're not able to authenticate silently.

As for refresh token expiration - AFAIK there is no way to inspect refresh token. It's not a JWT, like access token is, and there is nothing about refresh token parsing in client code, so I suspect that only server knows how to deal with and validate them.

See also [this](http://www.cloudidentity.com/blog/2015/03/20/azure-ad-token-lifetime) for the reasons of refresh tokens expiration:
> Refresh tokens can be invalidated at **_ANY_** time, for reasons independent from your app (e.g. user changes password). Hence you should NOT take a dependency on the above in your code – your logic should always assume that the refresh token can fail at any time. Refresh tokens issues for guest MSA accounts _**last only 12 hours**_

**Comment by vladimir-kotikov** [comment]
@tumetus, were you able to find the cause of your token expiration problem?

**Comment by tumetus** [comment]
@vladimir-kotikov Thanks for your response! Unfortunatelly I haven't yet had time to investigate this further.

**Comment by matteobortolazzo** [comment]
Did someone manage to make acquireTokenSilent() working after access token expiration?
